File: 01_Language/01_Functions/python/array_filter.py
# coding: utf-8

def array_filter(arr, cb=None):
    return filter(cb, arr)


if __name__ == '__main__':
    array1 = {'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5}
    array2 = [6, 7, 8, 9, 10, 11, 12]


    def odd(e):
        _, v = e
        return v % 2 == 1


    def even(e):
        return e % 2 == 0


    print(list(array_filter(array1.items(), odd)))
    print(dict(array_filter(array1.items(), odd)))
    print(list(array_filter(array2, even)))

    entry = ['foo', False, -1, None, '', '0', 0]
    print(list(array_filter(entry)))

File: 01_Language/01_Functions/js/array_filter.js
function array_filter(arr, cb) {
    cb = cb || function (e) {
        return !!e
    }
    let res = Array.isArray(arr) ? [] : {}
    for (let i in arr) {
        if (cb(arr[i])) {
            res[i] = arr[i]
        }
    }
    return res
}

let array1 = {'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5}
let array2 = [6, 7, 8, 9, 10, 11, 12]

console.log(array_filter(array1, function (e) {
    return e % 2 === 1
}));
console.log(array_filter(array2, function (e) {
    return e % 2 === 0
}));
